In [1]:
import pandas as pd

### Let's see how to import data from CSV, API(JSON), SQL Database...etc and hold the data in a dataframe object

In [2]:
df_movies = pd.read_csv("IMDb movies.csv")

##### Display a few rows first using head method

In [3]:
df_movies.head()

imdb_title_id                                              title  \
0     tt0000574                        The Story of the Kelly Gang   
1     tt0001892                                     Den sorte drøm   
2     tt0002101                                          Cleopatra   
3     tt0002130                                          L'Inferno   
4     tt0002199  From the Manger to the Cross; or, Jesus of Naz...   

                                      original_title  year date_published  \
0                        The Story of the Kelly Gang  1906     1906-12-26   
1                                     Den sorte drøm  1911     1911-08-19   
2                                          Cleopatra  1912     1912-11-13   
3                                          L'Inferno  1911     1911-03-06   
4  From the Manger to the Cross; or, Jesus of Naz...  1912           1913   

                       genre  duration           country language  \
0    Biography, Crime, Drama        70         Australia      NaN   
1                      Drama        53  Germany, Denmark      NaN   
2             Drama, History       100               USA  English   
3  Adventure, Drama, Fantasy        68             Italy  Italian   
4           Biography, Drama        60               USA  English   

                              director          ...           \
0                         Charles Tait          ...            
1                            Urban Gad          ...            
2                   Charles L. Gaskill          ...            
3  Francesco Bertolini, Adolfo Padovan          ...            
4                        Sidney Olcott          ...            

                                              actors  \
0  Elizabeth Tait, John Tait, Norman Campbell, Be...   
1  Asta Nielsen, Valdemar Psilander, Gunnar Helse...   
2  Helen Gardner, Pearl Sindelar, Miss Fielding, ...   
3  Salvatore Papa, Arturo Pirovano, Giuseppe de L...   
4  R. Henderson Bland, Percy Dyer, Gene Gauntier,...   

                                         description avg_vote votes   budget  \
0  True story of notorious Australian outlaw Ned ...      6.1   537   $ 2250   
1  Two men of high rank are both wooing the beaut...      5.9   171      NaN   
2  The fabled queen of Egypt's affair with Roman ...      5.2   420  $ 45000   
3  Loosely adapted from Dante's Divine Comedy and...      7.0  2019      NaN   
4  An account of the life of Jesus Christ, based ...      5.7   438      NaN   

   usa_gross_income worlwide_gross_income metascore reviews_from_users  \
0               NaN                   NaN       NaN                7.0   
1               NaN                   NaN       NaN                4.0   
2               NaN                   NaN       NaN               24.0   
3               NaN                   NaN       NaN               28.0   
4               NaN                   NaN       NaN               12.0   

   reviews_from_critics  
0                   7.0  
1                   2.0  
2                   3.0  
3                  14.0  
4                   5.0  

[5 rows x 22 columns]

#### Get a row using loc method and pass the index of the row

In [4]:
df_movies.loc[4]

imdb_title_id                                                    tt0002199
title                    From the Manger to the Cross; or, Jesus of Naz...
original_title           From the Manger to the Cross; or, Jesus of Naz...
year                                                                  1912
date_published                                                        1913
genre                                                     Biography, Drama
duration                                                                60
country                                                                USA
language                                                           English
director                                                     Sidney Olcott
writer                                                       Gene Gauntier
production_company                                           Kalem Company
actors                   R. Henderson Bland, Percy Dyer, Gene Gauntier,...
description              

#### Import new csv creating an index this time around using column imdb_title_id

__The imdb_title_id__ will no longer be a column in our data

In [5]:
df_movies = pd.read_csv("IMDb movies.csv", index_col='imdb_title_id')

In [6]:
df_movies.head(2)

title               original_title  year  \
imdb_title_id                                                                   
tt0000574      The Story of the Kelly Gang  The Story of the Kelly Gang  1906   
tt0001892                   Den sorte drøm               Den sorte drøm  1911   

              date_published                    genre  duration  \
imdb_title_id                                                     
tt0000574         1906-12-26  Biography, Crime, Drama        70   
tt0001892         1911-08-19                    Drama        53   

                        country language      director  \
imdb_title_id                                            
tt0000574             Australia      NaN  Charles Tait   
tt0001892      Germany, Denmark      NaN     Urban Gad   

                                              writer          ...           \
imdb_title_id                                                 ...            
tt0000574                               Charles Tait          ...            
tt0001892      Urban Gad, Gebhard Schätzler-Perasini          ...            

                                                          actors  \
imdb_title_id                                                      
tt0000574      Elizabeth Tait, John Tait, Norman Campbell, Be...   
tt0001892      Asta Nielsen, Valdemar Psilander, Gunnar Helse...   

                                                     description avg_vote  \
imdb_title_id                                                               
tt0000574      True story of notorious Australian outlaw Ned ...      6.1   
tt0001892      Two men of high rank are both wooing the beaut...      5.9   

               votes  budget usa_gross_income worlwide_gross_income metascore  \
imdb_title_id                                                                   
tt0000574        537  $ 2250              NaN                   NaN       NaN   
tt0001892        171     NaN              NaN                   NaN       NaN   

               reviews_from_users  reviews_from_critics  
imdb_title_id                                            
tt0000574                     7.0                   7.0  
tt0001892                     4.0                   2.0  

[2 rows x 21 columns]

In [7]:
df_test = pd.read_csv("IMDb movies.csv")

In [8]:
df_test.head(1)

imdb_title_id                        title               original_title  \
0     tt0000574  The Story of the Kelly Gang  The Story of the Kelly Gang   

   year date_published                    genre  duration    country language  \
0  1906     1906-12-26  Biography, Crime, Drama        70  Australia      NaN   

       director          ...           \
0  Charles Tait          ...            

                                              actors  \
0  Elizabeth Tait, John Tait, Norman Campbell, Be...   

                                         description avg_vote votes  budget  \
0  True story of notorious Australian outlaw Ned ...      6.1   537  $ 2250   

   usa_gross_income worlwide_gross_income metascore reviews_from_users  \
0               NaN                   NaN       NaN                7.0   

   reviews_from_critics  
0                   7.0  

[1 rows x 22 columns]

### Lets confirm if the imdb_title_id is really unique for every row by using the count function in pandas and storing in a set, then compare

In [9]:
x = df_test['imdb_title_id']

In [10]:
x.count() == len(set(x))

True

In [11]:
df_movies.head(1)

title               original_title  year  \
imdb_title_id                                                                   
tt0000574      The Story of the Kelly Gang  The Story of the Kelly Gang  1906   

              date_published                    genre  duration    country  \
imdb_title_id                                                                
tt0000574         1906-12-26  Biography, Crime, Drama        70  Australia   

              language      director        writer          ...           \
imdb_title_id                                               ...            
tt0000574          NaN  Charles Tait  Charles Tait          ...            

                                                          actors  \
imdb_title_id                                                      
tt0000574      Elizabeth Tait, John Tait, Norman Campbell, Be...   

                                                     description avg_vote  \
imdb_title_id                                                               
tt0000574      True story of notorious Australian outlaw Ned ...      6.1   

               votes  budget usa_gross_income worlwide_gross_income metascore  \
imdb_title_id                                                                   
tt0000574        537  $ 2250              NaN                   NaN       NaN   

               reviews_from_users  reviews_from_critics  
imdb_title_id                                            
tt0000574                     7.0                   7.0  

[1 rows x 21 columns]

## Importing Data from a REST API (i.e JSON)

_In python we decode JSON into dictionaries_

In [12]:
df_posts = pd.read_json("https://jsonplaceholder.typicode.com/posts")

_Check if any column returns true i.e has unique values for each row_

In [13]:
df_posts['id'].count() == len(set(df_posts['id']))

True

In [14]:
df_posts.head(1)

body  id  \
0  quia et suscipit\nsuscipit recusandae consequu...   1   

                                               title  userId  
0  sunt aut facere repellat provident occaecati e...       1

### The read_json method doesnt have the index_col arguement, we therefore use set_index

In [15]:
df_posts.set_index('id', inplace=True)

In [16]:
df_posts.head()

body  \
id                                                      
1   quia et suscipit\nsuscipit recusandae consequu...   
2   est rerum tempore vitae\nsequi sint nihil repr...   
3   et iusto sed quo iure\nvoluptatem occaecati om...   
4   ullam et saepe reiciendis voluptatem adipisci\...   
5   repudiandae veniam quaerat sunt sed\nalias aut...   

                                                title  userId  
id                                                             
1   sunt aut facere repellat provident occaecati e...       1  
2                                        qui est esse       1  
3   ea molestias quasi exercitationem repellat qui...       1  
4                                eum et est occaecati       1  
5                                  nesciunt quas odio       1

In [17]:
df_movies.head(2)

title               original_title  year  \
imdb_title_id                                                                   
tt0000574      The Story of the Kelly Gang  The Story of the Kelly Gang  1906   
tt0001892                   Den sorte drøm               Den sorte drøm  1911   

              date_published                    genre  duration  \
imdb_title_id                                                     
tt0000574         1906-12-26  Biography, Crime, Drama        70   
tt0001892         1911-08-19                    Drama        53   

                        country language      director  \
imdb_title_id                                            
tt0000574             Australia      NaN  Charles Tait   
tt0001892      Germany, Denmark      NaN     Urban Gad   

                                              writer          ...           \
imdb_title_id                                                 ...            
tt0000574                               Charles Tait          ...            
tt0001892      Urban Gad, Gebhard Schätzler-Perasini          ...            

                                                          actors  \
imdb_title_id                                                      
tt0000574      Elizabeth Tait, John Tait, Norman Campbell, Be...   
tt0001892      Asta Nielsen, Valdemar Psilander, Gunnar Helse...   

                                                     description avg_vote  \
imdb_title_id                                                               
tt0000574      True story of notorious Australian outlaw Ned ...      6.1   
tt0001892      Two men of high rank are both wooing the beaut...      5.9   

               votes  budget usa_gross_income worlwide_gross_income metascore  \
imdb_title_id                                                                   
tt0000574        537  $ 2250              NaN                   NaN       NaN   
tt0001892        171     NaN              NaN                   NaN       NaN   

               reviews_from_users  reviews_from_critics  
imdb_title_id                                            
tt0000574                     7.0                   7.0  
tt0001892                     4.0                   2.0  

[2 rows x 21 columns]

### Pre-requisites of EDA (Exploratory Data Analysis) i.e Descriptive statics. Data preparation/cleaning steps are;

* Run df.info() to check column/row frequencies of null/non-noll values. 
* If more than half of the rows in a column has None/NA/Null values then we could drop the column.
* If the null values are few, then we can replace them with mean/median/mode.
* Deal with duplicate values. Remove and keep only one.

### Performing Descriptive Statistics

* Also run describe() method on your df. This will show frequency/mean/min/max/standard deviation/variance of Continous values(numbers only).
* Check the relationship between continous variables i.e one column and another

In [18]:
df_movies.info()

<class 'pandas.core.frame.DataFrame'>
Index: 81273 entries, tt0000574 to tt9914286
Data columns (total 21 columns):
title                    81273 non-null object
original_title           81273 non-null object
year                     81273 non-null int64
date_published           81273 non-null object
genre                    81273 non-null object
duration                 81273 non-null int64
country                  81234 non-null object
language                 80518 non-null object
director                 81200 non-null object
writer                   79780 non-null object
production_company       76948 non-null object
actors                   81207 non-null object
description              78843 non-null object
avg_vote                 81273 non-null float64
votes                    81273 non-null int64
budget                   22804 non-null object
usa_gross_income         15094 non-null object
worlwide_gross_income    29892 non-null object
metascore                12722 non-null 

__Country seems to have 39 null values. How would we deal with this__?

We replace the values with ameasure of tenderncy since they are few. 
In this case we choose the mode(most common country) because the data type is object (string). 

_NB: IF say we needed the actors column in our objectives then we would have to remove all rows which do not have actors_

This would be the case if you are performing a qualitative analysis instead of quantitative analysis.

__Budget column has 22,804 out of 81273 (i.e almost 72%) null  values. How would we deal with scenario?__

We would drop the entire column since it contains too many nulll values

__How would we deal with reviews_from_users?__

We would replace the null values with the mean since the data types is integer.

### An example of creating a dataframe from an SQL Table

In [19]:
import psycopg2 as ps

In [20]:
con = ps.connect("host=localhost user=postgres password=123456 dbname=sales_data")

In [21]:
df_sales =  pd.read_sql_query("select * from sales",con)

In [22]:
df_sales

id     name       sales
0   1  Colgate  2679387.45
1   2    Bread  7937347.45

### 1. Clean up our columns and check for duplicates

In [80]:
df_movies_one = pd.read_csv("data/imdb_data_1000.csv", index_col='Rank')

In [82]:
df_movies_one.columns

Index(['Title', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime (Minutes)', 'Rating', 'Votes', 'Revenue (Millions)',
       'Metascore'],
      dtype='object')

In [83]:
df_movies_one.rename(columns = {'Runtime (Minutes)' : 'Runtime_Minutes',
                               'Revenue (Millions)': 'Revenue_Millions'}, 
                                 inplace=True)

In [78]:
df_movies_one.columns = [c.lower() for c in df_movies_one]

In [79]:
df_movies_one.columns

Index(['title', 'genre', 'description', 'director', 'actors', 'year',
       'runtime_minutes', 'rating', 'votes', 'revenue_millions', 'metascore'],
      dtype='object')

In [45]:
df_movies_one.shape

(1000, 11)

In [49]:
df_movies_one.drop_duplicates(keep='first', inplace=True)

In [50]:
df_movies_one.shape

(1000, 11)

#### Create a temporary data frame, create duplicate values using append method

In [51]:
tmp_df = df_movies_one.append(df_movies_one)

In [53]:
tmp_df = tmp_df.append(df_movies_one)

In [54]:
tmp_df.shape

(3000, 11)

In [55]:
tmp_df_one = tmp_df.copy()

In [58]:
tmp_df.drop_duplicates(inplace=True,keep='last')

In [61]:
tmp_df.shape

(1000, 11)

In [63]:
tmp_df_one.drop_duplicates(inplace=True, keep= False)

In [64]:
tmp_df_one.shape

(0, 11)

In [66]:
df_movies.shape

(81273, 21)

In [67]:
df_movies_one.shape

(1000, 11)

In [68]:
df_comb = df_movies_one.append(df_movies)

C:\Users\edwin\Anaconda3\lib\site-packages\pandas\core\frame.py:6211: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  sort=sort)


In [70]:
df_comb.shape

(82273, 24)

In [75]:
df_comb.columns

Index(['actors', 'avg_vote', 'budget', 'country', 'date_published',
       'description', 'director', 'duration', 'genre', 'language', 'metascore',
       'original_title', 'production_company', 'rating', 'revenue_millions',
       'reviews_from_critics', 'reviews_from_users', 'runtime_minutes',
       'title', 'usa_gross_income', 'votes', 'worlwide_gross_income', 'writer',
       'year'],
      dtype='object')

In [76]:
df_movies.columns

Index(['title', 'original_title', 'year', 'date_published', 'genre',
       'duration', 'country', 'language', 'director', 'writer',
       'production_company', 'actors', 'description', 'avg_vote', 'votes',
       'budget', 'usa_gross_income', 'worlwide_gross_income', 'metascore',
       'reviews_from_users', 'reviews_from_critics'],
      dtype='object')

In [77]:
df_movies_one.columns

Index(['title', 'genre', 'description', 'director', 'actors', 'year',
       'runtime_minutes', 'rating', 'votes', 'revenue_millions', 'metascore'],
      dtype='object')